File: check_braces.js

const fs = require('fs');

const filename = "c:\\Users\\DELL\\Downloads\\portel-master (2)\\portel-master\\client\\index.html";
const content = fs.readFileSync(filename, 'utf8');

let stack = [];
const pairs = { ')': '(', '}': '{', ']': '[' };

let line_no = 1;
let col_no = 1;

let in_string = false;
let quote_char = '';
let in_comment = false;
let comment_type = ''; // 'single' or 'multi'

for (let i = 0; i < content.length; i++) {
    const char = content[i];
    const prev_char = i > 0 ? content[i - 1] : '';
    const next_char = i < content.length - 1 ? content[i + 1] : '';

    if (char === '\n') {
        line_no++;
        col_no = 1;
    } else {
        col_no++;
    }

    if (!in_comment && !in_string) {
        if (char === "'" || char === '"' || char === '`') {
            in_string = true;
            quote_char = char;
        } else if (char === '/' && next_char === '/') {
            in_comment = true;
            comment_type = 'single';
        } else if (char === '/' && next_char === '*') {
            in_comment = true;
            comment_type = 'multi';
        } else if ('({['.includes(char)) {
            stack.push({ char, line: line_no, col: col_no });
        } else if (')}]'.includes(char)) {
            if (stack.length === 0) {
                console.log(`Extra closing ${char} at line ${line_no}, col ${col_no}`);
            } else {
                const top = stack.pop();
                if (top.char !== pairs[char]) {
                    console.log(`Mismatched ${char} at line ${line_no}, col ${col_no}. Expected closure for ${top.char} from line ${top.line}, col ${top.col}`);
                }
            }
        }
    } else if (in_string) {
        if (char === quote_char && prev_char !== '\\') {
            in_string = false;
        }
    } else if (in_comment) {
        if (comment_type === 'single' && char === '\n') {
            in_comment = false;
        } else if (comment_type === 'multi' && char === '*' && next_char === '/') {
            // next iter will handle the /
        } else if (comment_type === 'multi' && prev_char === '*' && char === '/') {
            in_comment = false;
        }
    }
}

if (stack.length > 0) {
    console.log(`Found ${stack.length} unclosed symbols:`);
    stack.forEach(s => {
        console.log(`  Unclosed ${s.char} from line ${s.line}, col ${s.col}`);
    });
} else {
    console.log("All symbols are balanced!");
}

File: check_braces.py

import os

def check_balance(filename):
    if not os.path.exists(filename):
        print(f"File not found: {filename}")
        return
        
    with open(filename, 'r', encoding='utf-8') as f:
        content = f.read()

    stack = []
    pairs = {')': '(', '}': '{', ']': '['}
    
    line_no = 1
    col_no = 1
    
    in_string = False
    quote_char = ''
    in_comment = False
    comment_type = '' # 'single' or 'multi'
    
    for i, char in enumerate(content):
        # Progress trackers
        current_char = char
        prev_char = content[i-1] if i > 0 else ''
        next_char = content[i+1] if i < len(content)-1 else ''

        if char == '\n':
            line_no += 1
            col_no = 1
        else:
            col_no += 1
            
        if not in_comment and not in_string:
            if char in "'\"`":
                in_string = True
                quote_char = char
            elif char == '/' and next_char == '/':
                in_comment = True
                comment_type = 'single'
            elif char == '/' and next_char == '*':
                in_comment = True
                comment_type = 'multi'
            elif char in '({[':
                stack.append((char, line_no, col_no))
            elif char in ')}]':
                if not stack:
                    print(f"Extra closing {char} at line {line_no}, col {col_no}")
                else:
                    top, l, c = stack.pop()
                    if top != pairs[char]:
                        print(f"Mismatched {char} at line {line_no}, col {col_no}. Expected closure for {top} from line {l}, col {c}")
        elif in_string:
            if char == quote_char and prev_char != '\\':
                in_string = False
        elif in_comment:
            if comment_type == 'single' and char == '\n':
                in_comment = False
            elif comment_type == 'multi' and char == '*' and next_char == '/':
                # We'll skip the '/' in the next iteration implicitly by logic or just handle it
                pass 
            elif comment_type == 'multi' and prev_char == '*' and char == '/':
                in_comment = False

    if stack:
        print(f"Found {len(stack)} unclosed symbols:")
        for char, l, c in stack:
            print(f"  Unclosed {char} from line {l}, col {c}")
    else:
        print("All symbols are balanced!")

if __name__ == "__main__":
    check_balance("c:\\Users\\DELL\\Downloads\\portel-master (2)\\portel-master\\client\\index.html")
